File: android/src/main/java/com/lightningkite/khrysalis/rx/DisposeCondition.shared.kt
package com.lightningkite.khrysalis.rx

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.lightningkite.khrysalis.escaping
import com.lightningkite.khrysalis.observables.Closeable
import com.lightningkite.khrysalis.swiftDescendsFrom
import io.reactivex.disposables.Disposable

class DisposeCondition(val call: @escaping() (Disposable) -> Unit)

infix fun DisposeCondition.and(other: DisposeCondition): DisposeCondition =
    andAllDisposeConditions(listOf(this, other))

fun andAllDisposeConditions(list: List<DisposeCondition>): DisposeCondition = DisposeCondition { it ->
    var disposalsLeft = list.size
    for (item in list) {
        item.call(DisposableLambda {
            disposalsLeft -= 1
            if (disposalsLeft == 0) it.dispose()
        })
    }
}

infix fun DisposeCondition.or(other: DisposeCondition): DisposeCondition =
    DisposeCondition { it -> this.call(it); other.call(it) }

File: ios/Khrysalis/com/lightningkite/khrysalis/rx/DisposeCondition.shared.swift
//Package: com.lightningkite.khrysalis.rx
//Converted using Khrysalis2

import Foundation
import Khrysalis
import RxSwift
import RxRelay



public class DisposeCondition {
    
    public var call:  (Disposable) -> Void
    
    
    public init(call: @escaping (Disposable) -> Void) {
        self.call = call
    }
    convenience public init(_ call: @escaping (Disposable) -> Void) {
        self.init(call: call)
    }
}
 
 

extension DisposeCondition {
    public func and(other: DisposeCondition) -> DisposeCondition {
        return andAllDisposeConditions([self, other])
    }
    public func and(_ other: DisposeCondition) -> DisposeCondition {
        return and(other: other)
    }
}
 
 

public func andAllDisposeConditions(list: Array<DisposeCondition>) -> DisposeCondition {
    return DisposeCondition{ (it) in 
        var disposalsLeft = list.size
        
        for item in list {
            item.call(DisposableLambda{ () in 
                disposalsLeft -= 1
                if disposalsLeft == 0 {
                    it.dispose()
                }
            })
        }
    }
}
public func andAllDisposeConditions(_ list: Array<DisposeCondition>) -> DisposeCondition {
    return andAllDisposeConditions(list: list)
}
 
 

extension DisposeCondition {
    public func or(other: DisposeCondition) -> DisposeCondition {
        return DisposeCondition{ (it) in 
            self.call(it)
            other.call(it)
        }
    }
    public func or(_ other: DisposeCondition) -> DisposeCondition {
        return or(other: other)
    }
}
 
